### PR TITLE
[RNMobile] add selectionStart and selectionEnd to transientEdits

### PR DESCRIPTION
**Pull request by dratwas**
## Description
Gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2285

In this PR i added a `selectionStart` and `selectionEnd` to transient Edits. The motivation is described here: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2192#issuecomment-630010164

I also added a debounce function to create an undo level when the input (rich-text) stops for over a second to be consistent with the web: 
https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/component/index.js#L501

It changes how the undo mechanism works on mobile.

## Known Issues:
**These issues are not introduced by this PR**
- Sometimes the redo level is removed in rich-text - After testing, I can say that this PR doesn't introduce that issue but it is easier to reproduce. (GIFS in the testing section) I found that it happens only when the paragraph block has the break line at the very beginning. It's really edge case because as far as I know, it is possible by adding it to the code editor.

- We create unnecessary undo levels on the editor init - https://github.com/WordPress/gutenberg/pull/22492#issuecomment-631977832

## How has this been tested?

**Undo/redo block actions**

- From the example app, remove a few blocks.
- Press the Undo button to see them reappear.
- Press Redo to remove the blocks again.

**Undo/redo text**
NOTE: This test case doesn't work as before. The whole text is removed since we do not create the undo level when we change the same attribute in the same block. Before changes from this PR the undo level was created after each sign in the rich-text. After these changes, the undo works the same as on the web.

Before | After
--- | ---
undobefore | undoafter

Web | 
---|
undowebpar

- Write some text on a text based block .
- Press Undo until all new text has disappeared.
- Press Redo to see the text appear again.

**Edit: I observed a weird behavior. Sometimes when i remove text by pressing on the undo button the redo level is not created - see the gif below (the redo button is disabled after pressing undo)**
Issue | 
---|
paraissue

**Undo/redo text format**
In that case, i have the same issue as in the previous one - the redo level is removed in rich-text sometimes
Issue | 
---|
boldissue

- On a rich-text based component, add some format (bold, links, etc…).
- Undo all changes and Redo them to arrive to the same state.

**I've tested that and it seems like these issues i mentioned above also appear w/o changes from this PR**

## Types of changes
Add `selectionStart` and `selectionEnd` to transientEdits to make undo mechanism work the same as on the web

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by dratwas** [comment]
I thought that these changes (selectionStart and selectionEnd) are a cause of these issues but no, they are just easier to reproduce but appears also in the current state of the app. During the investigation, I found that the undo level (or few of them) is created on the init of the editor which I think is a bug. An interesting thing is that if the post/page doesn't contain some InnerBlocks then only one undo level is created and pressing on the undo button removes all blocks. In case you have i.e Cover block, then two undo levels are created and first remove the inner blocks from the Cover block and then remove all blocks.

No inner blocks | With inner blocks
---|---
undoinit | initbig

File: packages/rich-text/src/component/index.native.js [hunk 1]
```diff
@@ -88,7 +88,7 @@ export class RichText extends Component {
 		this.formatToValue = memize( this.formatToValue.bind( this ), {
 			maxSize: 1,
 		} );
-
+		this.debounceCreateUndoLevel = debounce( this.onCreateUndoLevel, 1000 );
```
Comment (mkevins) [inline packages/rich-text/src/component/index.native.js:91]: I see here that we are debouncing this for every invocation, but web seems to have some cases where the `setTimeout` is not used, e.g. in `onChange` they call `this.onCreateUndoLevel` directly: https://github.com/WordPress/gutenberg/blob/33bace57b13aa307f8befb609d0798a6a16dbfaf/packages/rich-text/src/component/index.js#L669

I wonder, are there reasons for not debouncing on web in some cases? :thinking:

**Review by mkevins — APPROVED** [review]
I tested this on Pixel 3a and this PR works as described. I confirmed too that the issues mentioned are present in `develop`, and not introduced in this PR. I also observed another issue (which is also present on `develop`) where the caret did not "land" in the correct place after undo and redo actions:



Nice work on this PR @dratwas , and thanks for documenting your findings with great detail! :smile:

File: packages/rich-text/src/component/index.native.js (hunk 1 above)
Comment (dratwas) [reply packages/rich-text/src/component/index.native.js:91]: We also have one case when we don't use debounce - onFormatChange which is also called in `onInsert`. I haven't changed the `onCreateUndoLevel` itself but i created a new function that is called `debounceCreateUndoLevel` and invoke it on each input change as they do here :) https://github.com/WordPress/gutenberg/blob/33bace57b13aa307f8befb609d0798a6a16dbfaf/packages/rich-text/src/component/index.js#L501

As far as i can see they also call `onChage` inside `onInput` but with `withoutHistory ` set to true which turns off creating of undo level. https://github.com/WordPress/gutenberg/blob/33bace57b13aa307f8befb609d0798a6a16dbfaf/packages/rich-text/src/component/index.js#L488

Comment (mkevins) [reply packages/rich-text/src/component/index.native.js:91]: Ah, right! Thanks for the response! I misread the constructor bind as a same-named function :man_facepalming: . Thanks for the explanation, and great work on improving this!

**Review by chipsnyder — APPROVED** [review]
Tested on iOS iPhone 11 works as described for me as well 👍

**Comment by mchowning** [comment]
:wave: It looks like this change has introduced a regression with respect to saving a post when someone is typing quickly: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2349.

**Pull request merged**